### ISSUE TITLE
Growing trend in producer to consumer latency - is it normal?

### ISSUE THREAD

**Issue by lukasz-gawron** [aws/amazon-kinesis-video-streams-parser-library#82]
**Our setup**
**Producer**
Built from two Gstreamer pipelines.

`filesrc location=/root/bin/producer/video/movie.mp4 ! qtdemux ! video/x-h264,format=avc,alignment=au ! h264parse ! video/x-h264,stream-format=avc,alignment=au ! appsink name=H264_SINK caps=video/x-h264,stream-format=avc,alignment=au`

and

`appsrc name=video_src caps=video/x-h264,stream-format=avc,alignment=au ! queue ! video/x-h264,stream-format=avc,alignment=au ! kvssink stream-name="StreamName" credential-path=\"/root/.aws/kinesis_credentials\" aws-region=\"eu-west-1\" absolute-fragment-times=true rotation-period=70 `

Video frames are slightly modified in appsink callback (attached producer system time), and pushed to appsrc which is than pushed to Kinesis through kvssink.

**Kinesis stream**
Stream have retention period set to 1h.

**Consumer**
Consumer is based on example:
https://github.com/aws/amazon-kinesis-video-streams-parser-library/blob/master/src/main/java/com/amazonaws/kinesisvideo/parser/examples/KinesisVideoGStreamerPiperExample.java

But we've done few modifications, most outstanding are:
https://github.com/aws/amazon-kinesis-video-streams-parser-library/blob/2e93ba3ec843677c3128a6b4355e67f340ef0859/src/main/java/com/amazonaws/kinesisvideo/parser/examples/KinesisVideoGStreamerPiperExample.java#L125

changed to 100000 seconds.

And SelectorType changed from EARLIEST to NOW.

https://github.com/aws/amazon-kinesis-video-streams-parser-library/blob/2e93ba3ec843677c3128a6b4355e67f340ef0859/src/main/java/com/amazonaws/kinesisvideo/parser/examples/KinesisVideoGStreamerPiperExample.java#L105

Consumer architecture is like on image below:
<img width="775" alt="obraz" src="https://user-images.githubusercontent.com/3419338/75244157-9354b080-57cb-11ea-92e7-66383c6f9a97.png">

**PROBLEM**
We observed growing latency trend on our consumer. We will look more closely to efficiency of our producer and consumer pipelines but based on checks we've done till now seems that cpu and memory of consumer looks ok.

![obraz](https://user-images.githubusercontent.com/3419338/75244652-79679d80-57cc-11ea-9f96-8ef0339f0123.png)

Every 45 mins we may see that there is drop in latency. Its caused by some automatic process on consumer which refresh a session? (probably handled by https://github.com/aws/amazon-kinesis-video-streams-parser-library/blob/master/src/main/java/com/amazonaws/kinesisvideo/parser/examples/ContinuousGetMediaWorker.java). Latency drops a bit but trend is still growing.
**We expect this line to be more or less flat.  Is that correct?**

**Expectation**
1. Clarification is this growing trend normal or its something wrong?
2. Advices where we may look to identify root cause of that, if something is wrong with our setup.

Best regards,
Lukasz

**Comment by MushMal** [comment]
Anyone is looking at this? I wonder how they measure the latency – can’t really understand from the description.

**Comment by MushMal** [comment]
@lukasz-gawron this is an awesome post!

Could you please clarify how you measure the latency? 

While it's not immediately clear to me, it seems that you are attempting to combine two pipelines where the ultimate source is a file source. When using file source, the KVS GStreamer plugin will stream in "offline" mode. This means that it will fill in the producer buffer at the file I/O speed and block until there is availability in the streaming buffer while the networking streams out the bits. This will maximize the existing bandwidth/throughput to stream out as fast as it actually can.

You are also using file auth integration but you are specifying rotation-period which is not needed. The file based credential provider boils down to the following API call:

https://github.com/awslabs/amazon-kinesis-video-streams-producer-sdk-cpp/blob/0888c666efbe7f20891a86c9e64a677f73e33b17/kinesis-video-c-producer/src/source/Common/FileCredentialProvider.c#L116

to load the credentials from the specified file. NOTE: if you are using session token then it will attempt to load the expiration. Max expiration duration is 40 minutes on the producer side. At the session expiration a new session is established which is a new PutMedia call and it actually is an entirely different streaming session into the same logical stream.

As the streaming is happening at the max throughput speed it's interesting to see the producer side logs to ensure there are no issues on the producer side and we can see what's the observed bandwidth and whether there is throttling that's happening.

**Comment by lukasz-gawron** [comment]
Hi @MushMal,
thanks for feedback.

We are measuring latency in following way:
1. sync producer and consumer machines to the same ntp server.
2. [producer] appsink callback on pipeline 1 will overwrite last bytes of encoded h264 packet from file with current system time of producer machine, send that to appsrc pipeline nr 2
3. [producer] appsrc will forward modified samples to kvssink 
(**our intention would be to respect framerate from file which is 25 frames /sec. We are not setting streaming-type on kvssink, so it should be default realtime, not offline you mentioned - but not sure is this realtime mode guarantee respecting framerate**) From appsink -> appsrc we are pushing it that way: `GstFlowReturn ret = gst_app_src_push_sample(appSrc, newSample);`
4. [consumer java] java consumer reads frames from Kinesis, and pass it to CPP GStreamer application 
5. [consumer cpp] running pipeline `fdsrc ! matroskademux ! h264parse ! appsink`. On appsink callback we are substracting producer system time available in last bytes of h264  packet with consumer system time


After your previous comment I've checked what kvssink is logging every key frame and I'm confused please take a look:

**First key frame**
> [DEBUG] [14-02-2020 11:02:13:642.840 GMT] Kinesis Video client and stream metrics
	>> Overall storage byte size: 134217728
	>> Available storage byte size: 133873267
	>> Allocated storage byte size: 344461
	>> Total view allocation byte size: 144072
	>> Total streams frame rate (fps): 25
	>> Total streams transfer rate (bps): 33554432 (32768 Kbps)
	>> Current view duration (ms): 0
	>> Overall view duration (ms): 0
	>> Current view byte size: 344412
	>> Overall view byte size: 344412
	>> Current frame rate (fps): 25
	>> Current transfer rate (bps): 33554432 (32768 Kbps)

**Second key frame**
[DEBUG] [14-02-2020 11:02:14:103.406 GMT] Kinesis Video client and stream metrics
	>> Overall storage byte size: 134217728
	>> Available storage byte size: 133412008
	>> Allocated storage byte size: 805720
	>> Total view allocation byte size: 144072
	>> Total streams frame rate (fps): 5150
	>> Total streams transfer rate (bps): 31876704 (31129 Kbps)
	>> Current view duration (ms): 80
	>> Overall view duration (ms): 480
	>> Current view byte size: 369443
	>> Overall view byte size: 805034
	>> Current frame rate (fps): 5150.42
	>> Current transfer rate (bps): 31876704 (31129 Kbps)

**In second log we've got 5150.42**!! 
I've compared that with our logs which are printed on producer app sink and just before log from kvssink we've got:
2020-02-14 11:02:14.102,**20**,4348110377319217616,,344882,132261517341020000,LG

where second column with value 20 is frame count incremented with every frame processed in producer app sink callback.
**So we read 20 frames till now but kvssink printed more than 5k fps.**

Not sure how to reason about this, as I have multiple such strange values in log file. Max value is 18k fps.

Based on all values logged in  "Current frame rate (fps):" line  I've done this chart:
![image](https://user-images.githubusercontent.com/3419338/75363712-caee5600-58ba-11ea-8e0e-f5ef4fcd848a.png)

In files we have 230133 frames, and 18697 key frames (counted that based on number of kvssink logs)

Thats all I was able to dig out today to help describe  issues. I will work on that tomorrow so hopefully will be able to provide more details.

Thanks for your support,
Lukasz

**Comment by MushMal** [comment]
Thanks @lukasz-gawron 

The frame rate is calculation is based not only on the number of frames but also the duration at which those have been received. The calculations are then passed through an EMA (Exponential Mean Averaging) to add "inertia" to the graph. In your case, your pipeline is "bursty" and seems to be producing a lot of frames in short period of time, causing this metrics skew. The same seem to be with the transfer rate. The observed rate is around 32Mbps.

This said, I am still not entirely sure how this causes the end-to-end latency increases seen in your graph in the first post. From what I can tell from it, the latency continuously increases for around 26 seconds until it drops slightly, then it continues increasing in a similar pattern.

**Comment by lukasz-gawron** [comment]
Hi @MushMal ,
**in terms of 26 seconds** you mentioned, probably its confusing from our chart, but number 26k on X axis is actually 45 minutes and that correlates with automatic refresh of Java consumer I mentioned in first post. Unfortunately that how its measured - every 100 ms.

**In terms of current frame rate printed on kvssink** - thanks for explanation. We've analysed few other logs, and it seems that this value is wrong especially at beginning of kvssink work. First output is printed probably based on same stream caps. And next one is very high, and than it is impacting next rates as well.

**Would you be able to clarify what is the reason of latency drop at every 45 mins interval?** in chart in first post. why java consumer is restarting itself?

**Comment by MushMal** [comment]
@lukasz-gawron could you please check the CW metrics?

https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/monitoring-cloudwatch.html

Also, you can provide your account id or email it to kinesis-video-support@amazon.com

**Comment by bdhandap** [comment]
@lukasz-gawron Can you look at the `GetMedia.MillisBehindNow` metric in your cloudwatch? It should give the metric on how much the consumer is lagging behind the tip of the producer.

**Comment by sharadrj** [comment]
@lukasz-gawron, 

Can you share the producer-side log file with us? By default, it should be saved as kinesis-video-native-build/log/kvs.log. 

Also, can you confirm the target bit-rate and resolution.

**Comment by lukasz-gawron** [comment]
In meantime we’ve changed our producer pipeline and its:
appsrc name=mitch caps=video/x-h264,stream-format=avc,alignment=au ! queue  ! video/x-h264,stream-format=avc,alignment=au !  kvssink stream-name=\"" + streamName + "\" credential-path=\"/root/.aws/kinesis_credentials\"  aws-region=\"eu-west-1\" streaming-type=offline file-start-time=1581933141
 
 
Today (Monday 2nd) we repeated tests:
Our latency measurement is like this.
<img width="1316" alt="image" src="https://user-images.githubusercontent.com/3419338/75788527-a2f26d00-5d68-11ea-8564-63f39d851a0f.png">


Sorry for that this chart is confusing as we have 230k frames in total (2.5h) in our video file, but it shows just ~90k frames, as every datapoint is collected every 100 ms. I’ve also forget to sync time on producer before test start, so I got latency negative, but growing trend is clearly visible.
**I’ve done this test on a stream on Ireland region. Producer is running in Katowice, Poland.**

Here is Cloudwatch chart for suggested in my post GetMedia.MillisBehindNow
<img width="1312" alt="image" src="https://user-images.githubusercontent.com/3419338/75788610-ca493a00-5d68-11ea-8a51-faa6133b64e3.png">

I read it as Java consumer is not having growing trend, so  other potential cause of growing trend is
producer
CPP consumer with GStreamer application which is calculating latency after java consumer pass video frames to that process
 
 
We’ve done other tests to narrow root cause.
Running the same test from ec2 instance -> kinesis -> ec2 instance resulted in this chart
X Axis – frame count
Y axis – latency in ms

<img width="1314" alt="image" src="https://user-images.githubusercontent.com/3419338/75788713-e947cc00-5d68-11ea-9f69-f88b49796af4.png">

We may see here few 45 minutes sections. Second section here is ideally flat line starting at ~26k frames which is 45 mins time. Its very strange.
We suspect that after each GetMediaResult refresh we are hitting different Kinesis instance and we get different latencies based on server we are connected during such session – not sure how to confirm that.
**But most important thing in this test is it doesn’t have  growing trend which suggests that network connection from Poland to Ireland is major factor in this growing trend.**

To prove that we’ve done another test when we simulated slow network when producer was setup in AWS and we were able to reproduce growing trend:
<img width="1304" alt="image" src="https://user-images.githubusercontent.com/3419338/75788896-2b710d80-5d69-11ea-86cc-85520572499d.png">


On red are latencies added on eth0 interface by Linux command.
`sudo tc qdisc add dev eth0 root netem delay 500ms`
 
Tomorrow we will do a test with stream and consumer in Frankfurt instead of Ireland, as this regions is closer to us. We expect to have at least smaller angle of growing trend or ideally flat line as test where all components are in AWS.
 
Would be good to hear from you what other things we may check/tune to find and remove this growing latency.

Thanks,
Lukasz

**Comment by lukasz-gawron** [comment]
@sharadrj producer logs send to kinesis-video-support@amazon.com
Please let me know whether you receive it.

We've done a tests today for Frankfurt region but it doesn't help. Growing trend is still there.

We are going to look at producer performance and will try to find any metric in CloudWatch which measure time between frames left kvssink and reached Kinesis. Is there such metric?

Thanks,
Lukasz

**Comment by lukasz-gawron** [comment]
Target bit-rate is 8Mbit/s and resolution is 3840 x 2160

**Comment by MushMal** [comment]
@lukasz-gawron thanks for the logs. I have looked through and here are some observations.

Your entire stream has no errors whatsoever. There are 6 streaming sessions. The observed/required throughput is around 16Mbps (though you seem to indicate that the stream itself is 8Mbps). 

There are 5-6 instances where due to network congestion/limited throughput dipping to around 8 Mbps your buffer accumulates up-to 4 seconds of data but it lasts for around 2-3 seconds most and the buffer gets depleted once the throughput is back.

Looking at the producer side logs I can state that there are no delays. Now, we need to see what's wrong with the consumer/GetMedia/Gstreamer pipeline side.


Details:

kvs.log.2020-03-02-10

Stream name: MitchStream

Starts at:    02-03-2020 10:30:51:907.821 GMT
Ends at:      02-03-2020 10:59:59:997.075 GMT

Observation: Regular stable key-frames (fragments) every 500ms which is too little. Recommend using 1sec+ fragmentation

The entire streaming is in one session. No errors. Nomimal throughput is 16Mbps. Three ocasions when the throughput dips below to 8-12 Mbps and the buffer starts accumulating to around 3.6 seconds. The throughput gets back to above nominal allowing the stream to catch up. Most of the streaming has latency of 0 - 400ms with Persisted ACKs coming back within 500-600M (fantastic!).

Overall content store size: 134217728 bytes
Utilized size peak during streaming: 6778997 bytes

No issues can be seen.


====================================

kvs.log.2020-03-02-11

Stream name: MitchStream

Starts at:    02-03-2020 11:00:00:031.034 GMT
Ends at:      02-03-2020 12:00:00:066.584 GMT


Seems the continuation of the previous streaming session. First session terminates and the second starts with no hickups and gentle switch over to the new streaming session with 0 duration in the buffer

[INFO ] [02-03-2020 11:08:57:347.164 GMT] postReadCallback(): Reported end-of-stream for stream MitchStream. Upload handle: 0

The pattern is similar and we have almost no delay in the buffer.

Session 2 ends gracefully and swaps over to session 3 with no buffered frames:

[INFO ] [02-03-2020 11:10:49:973.464 GMT] getStreamData(): Indicating an EOS after last persisted ACK is received for stream upload handle 1


The streaming is OK with single instance when the throughput drops 

[DEBUG] [02-03-2020 11:23:14:607.523 GMT] Kinesis Video client and stream metrics
	>> Overall storage byte size: 134217728
	>> Available storage byte size: 126181774
	>> Allocated storage byte size: 8035954
	>> Total view allocation byte size: 144072
	>> Total streams frame rate (fps): 25
	>> Total streams transfer rate (bps): 10725360 (10473 Kbps)
	>> Current view duration (ms): 4280
	>> Overall view duration (ms): 5760

and we accumulate 4.2 seconds in the buffer with overall 5.7 seconds persisted ACK delay. A few seconds later the throughput is back to normal and we catch up to 0 delay.


There are few other hickups with the throughput leading to up-to 3 seconds in the buffer but as before, it catshes up within 2-3 seconds.


Session #2 ends and #3 starts - just like before with no delay:

[INFO ] [02-03-2020 11:47:47:231.167 GMT] postReadCallback(): Reported end-of-stream for stream MitchStream. Upload handle: 2

Session #3 ends and #4 starts - no delay:

[INFO ] [02-03-2020 11:50:47:546.205 GMT] postReadCallback(): Reported end-of-stream for stream MitchStream. Upload handle: 3



=========================================


kvs.log.2020-03-02-12

This is the continuation of the previous streaming session. Same pattern.

Switching to session #5 with no delay

[INFO ] [02-03-2020 12:28:13:517.166 GMT] postReadCallback(): Reported end-of-stream for stream MitchStream. Upload handle: 4


then, session #6 with 1.2 seconds in the buffer due to slow thgroughput but this catches up in a second.

[INFO ] [02-03-2020 12:30:45:504.375 GMT] postReadCallback(): Reported end-of-stream for stream MitchStream. Upload handle: 5


=======================================

kvs.log.2020-03-02-13

Continuing from the last session


Same pattern.

Log terminates at 02-03-2020 13:02:13:609.182 GMT

**Comment by MushMal** [comment]
@lukasz-gawron just to make sure we get some progress on this issue. Could you please ensure your pipeline produces over a second fragments and re-run the tests?

**Comment by lukasz-gawron** [comment]
@MushMal 
Thank you very much for your insights. 
Yes we will apply your suggestions and re run the tests. I'm not sure whether that would be this week.

Will update this ticket as soon as we will have results.

**Comment by MushMal** [comment]
@lukasz-gawron Any update on this issue?

**Comment by lukasz-gawron** [comment]
@MushMal not yet. Priorites changed a little bit recently on our side but we should get back to checking your suggestions in this or next week.

Just to let you know  in meantime our network conditions changed and we rewrited our GStreamer consumer application  (codebase was cleaned a lot, we also replaced app sink with pad probe on h264parse src pad to calculate latency), now we see flat latencies around 1 sec without changes on fragments range.

We didn't find root cause of previous issues, but we hope to do some tests to isolate every changed thing in meantime, so we will learn what was root cause.
